File: projects/gnomad-api/src/schema/datasets/exac/exacConstraint.js
import { GraphQLFloat, GraphQLInt, GraphQLObjectType } from 'graphql'
import { UserVisibleError } from '../../errors'

export const ExacConstraintType = new GraphQLObjectType({
  name: 'ExacConstraint',
  fields: {
    // Expected
    exp_syn: { type: GraphQLFloat },
    exp_mis: { type: GraphQLFloat },
    exp_lof: { type: GraphQLFloat },
    // Observed
    obs_syn: { type: GraphQLInt },
    obs_mis: { type: GraphQLInt },
    obs_lof: { type: GraphQLInt },
    // mu
    mu_syn: { type: GraphQLFloat },
    mu_mis: { type: GraphQLFloat },
    mu_lof: { type: GraphQLFloat },
    // Z
    syn_z: { type: GraphQLFloat },
    mis_z: { type: GraphQLFloat },
    lof_z: { type: GraphQLFloat },
    // Other
    pLI: { type: GraphQLFloat },
  },
})

export const fetchExacConstraintByTranscriptId = async (ctx, transcriptId) => {
  try {
    const response = await ctx.database.elastic.get({
      index: 'exac_constraint',
      type: 'documents',
      id: transcriptId,
    })

    return response._source
  } catch (err) {
    if (err.message === 'Not Found') {
      throw new UserVisibleError('ExAC constraint not found')
    }
    throw err
  }
}

File: data-pipeline/src/data_pipeline/datasets/exac/exac_constraint.py
import hail as hl


def prepare_exac_constraint(path):
    ds = hl.import_table(path, force=True)
    ds = ds.repartition(32, shuffle=True)

    # Select relevant fields
    ds = ds.select(
        # Remove version number from transcript ID
        transcript_id=ds.transcript.split("\\.")[0],
        # Expected
        exp_syn=hl.float(ds.exp_syn),
        exp_mis=hl.float(ds.exp_mis),
        exp_lof=hl.float(ds.exp_lof),
        # Actual
        obs_syn=hl.int(ds.n_syn),
        obs_mis=hl.int(ds.n_mis),
        obs_lof=hl.int(ds.n_lof),
        # mu
        mu_syn=hl.float(ds.mu_syn),
        mu_mis=hl.float(ds.mu_mis),
        mu_lof=hl.float(ds.mu_lof),
        # Z
        syn_z=hl.float(ds.syn_z),
        mis_z=hl.float(ds.mis_z),
        lof_z=hl.float(ds.lof_z),
        # Other
        pli=hl.float(ds.pLI),
    )

    ds = ds.key_by("transcript_id")

    return ds
